### PR TITLE
Tinkerpop-1641 Kerberos authentication for gremlin-python

### PR DESCRIPTION
**Pull request by vtslab**
[x] mvn clean install -DskipTests && mvn clean install -Pglv-python
[x] mvn clean install -pl gremlin-server -DskipIntegrationTests=false
[x] docker/build.sh -t
[x] bin/process-docs.sh
[x] manual connection to ws://172.17.0.2:45940/gremlin on docker/gremlin-server.sh with Gremlin Console and Gremlin Python
[x] manual connection to ws://172.17.0.2:45941/gremlin on docker/gremlin-server.sh with Gremlin Console and Gremlin Python
[x] manual connection to ws://172.17.0.2:45942/gremlin on docker/gremlin-server.sh with Gremlin Console and Gremlin Python

This feature ([Kerberos for gremlin-python](https://issues.apache.org/jira/browse/TINKERPOP-1641)) turned out to have a pretty large footprint because of the need to make the KDC server available in both gremlin-test and the gremlin-server-test docker container. Three concerns regarding the review:      
1. I am not entirely sure whether the gremlin-test jar-with-dependencies is the cleanest way to make the KDC available in the Docker container.        
2. Because of the sheer complexity of Kerberos configs and error messages I thought it useful to document the transcripts of the manual connections made against the gremlin-server-test docker container. If reviewers feel the same, please state what is the best location for it (possibly not the current location docker/gremlin-server/docker-entrypoint.sh).     
3. This work originated from an earlier ticket [TINKERPOP-1566](https://issues.apache.org/jira/browse/TINKERPOP-1566). That one was also reviewed by Mike Adamson on request from Stephen Mallette. Although the current work does not change the java Kerberos part of the code, it might still be worthwhile to involve him.

### CODE REVIEW

**Comment by vtslab** [comment]
I noticed the travis build issue in gremlin-python and will come back to it later.

**Comment by vtslab** [comment]
Does anyone with knowledge of gremlin-dotnet see what is the matter? All earlier travis builds (red crosses above) only failed on gremlin-python, so if gremlin-dotnet builds are consistent, the current failure suggests the addition of the gcc and python3.6-dev packages would somehow interfere with dotnet. 
This PR does introduce one change in gremlin-dotnet, but this is only in the test-server-start.groovy script and should not be relevant to existing tests.
If no ideas pop up, I will revert and reapply the gcc and python3.6-dev changes later this week and see what happens.

**Comment by vtslab** [comment]
What is there to do to stop travis whining?
- gremlin-dotnet now builds without any changes with respect to the earlier failing build
- latest build now fails on network issues (Could not transfer artifact org.codehaus.groovy:groovy:jar:2.5.11 from/to google-maven-central (https://maven-central.storage-download.googleapis.com/maven2/))

I leave it at that and just wait until the PR review is complete (no hurry).

**Comment by spmallette** [comment]
Sorry, it took me a bit to get to this:

> This feature (Kerberos for gremlin-python) turned out to have a pretty large footprint because of the need to make the KDC server available in both gremlin-test and the gremlin-server-test docker container. 

yep - that spun out of control quickly :)

> I am not entirely sure whether the gremlin-test jar-with-dependencies is the cleanest way to make the KDC available in the Docker container.

That struck me as well in review. My initial reaction was that I'd like to keep `gremlin-test` isolated to Gremlin provider testing and to move re-usable test components somewhere else. I think we have other odds and ends like this KDC fixture. I'll need to think about that a bit more.

> Because of the sheer complexity of Kerberos configs and error messages I thought it useful to document the transcripts of the manual connections made against the gremlin-server-test docker container. If reviewers feel the same, please state what is the best location for it (possibly not the current location docker/gremlin-server/docker-entrypoint.sh).

I liked that bit of documentation actually. I was content with where you put it for now, but I will think so more on it.

> Does anyone with knowledge of gremlin-dotnet see what is the matter? All earlier travis builds (red crosses above) only failed on gremlin-python, so if gremlin-dotnet builds are consistent, the current failure suggests the addition of the gcc and python3.6-dev packages would somehow interfere with dotnet.
> If no ideas pop up, I will revert and reapply the gcc and python3.6-dev changes later this week and see what happens.
> What is there to do to stop travis whining?

Travis is flakey and I don't think it's our tests that cause the trouble. The environment fails for all sorts of reasons (e.g. internet connections die, VMs randomly timeout for no reason, etc). I've yet to see a pattern of failure that involves anything we can fix on our end unfortunately. I pretty much just look for clear test failures in Travis as things to worry about and just re-run tests as needed that failed to get a green run. I also mostly rely on my own local build. If i get a clean build with `mvn` and a good docker build locally it should be pretty solid.

Travis seems happier now so whatever was the matter seems resolved. There was a failure with spark as of your last commit but I assume that's nothing. I restarted it and seems happy now

I'll start thinking about the items I mentioned above to see what changes I might suggest. Thanks for the work on this.

**Comment by spmallette** [comment]
After thinking this through (sorry it took a while), I think that the way you have things is fine. There remains a bigger picture for the tests in 3.5.0 but I don't think we'll go and apply that here for this to merge. 

I do think the hardcoded "3.5.0-SNAPSHOT" in `docker-entrypoint.sh` needs to be dynamically referenced. 

I am having trouble building this locally for some reason. There seems to be some trouble with the python kerberos package:

```text
main:
     [copy] Copying 61 files to /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3
     [exec] Python 3.6.9
     [exec] 15.1.0
     [exec] Already using interpreter /usr/bin/python3
     [exec] Using base prefix '/usr'
     [exec] New python executable in /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/bin/python3
     [exec] Also creating executable in /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/bin/python
     [exec] Installing setuptools, pkg_resources, pip, wheel...done.
     [exec] Requirement already satisfied: wheel in ./env/lib/python3.6/site-packages (0.34.2)
     [exec] Collecting radish-bdd
     [exec]   Using cached radish_bdd-0.13.1-py2.py3-none-any.whl (164 kB)
     [exec] Collecting PyHamcrest
     [exec]   Using cached PyHamcrest-2.0.2-py3-none-any.whl (52 kB)
     [exec] Collecting aenum
     [exec]   Using cached aenum-2.2.3-py3-none-any.whl (40 kB)
     [exec] Collecting isodate
     [exec]   Using cached isodate-0.6.0-py2.py3-none-any.whl (45 kB)
     [exec] Collecting kerberos
     [exec]   Using cached kerberos-1.3.0.tar.gz (19 kB)
     [exec] Collecting six
     [exec]   Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
     [exec] Collecting humanize
     [exec]   Using cached humanize-2.4.0-py3-none-any.whl (62 kB)
     [exec] Processing /home/smallette/.cache/pip/wheels/9b/04/dd/7daf4150b6d9b12949298737de9431a324d4b797ffd63f526e/docopt-0.6.2-py2.py3-none-any.whl
     [exec] Collecting parse-type>0.4.0
     [exec]   Using cached parse_type-0.5.2-py2.py3-none-any.whl (32 kB)
     [exec] Collecting colorful>=0.3.11
     [exec]   Using cached colorful-0.5.4-py2.py3-none-any.whl (201 kB)
     [exec] Collecting tag-expressions>=1.0.0
     [exec]   Using cached tag_expressions-1.1.0-py3-none-any.whl (5.6 kB)
     [exec] Processing /home/smallette/.cache/pip/wheels/d6/d7/a5/ebb3499392c6897132acb7cc89cdd8506d53176d61910afc66/pysingleton-0.2.1-cp36-none-any.whl
     [exec] Processing /home/smallette/.cache/pip/wheels/36/9d/03/65b7d0b19b0f1946a3c75c4a467dd2f8069a3a6a80c9101946/parse-1.15.0-py3-none-any.whl
     [exec] Building wheels for collected packages: kerberos
     [exec]   Building wheel for kerberos (setup.py): started
     [exec]   ERROR: Command errored out with exit status 1:
     [exec]   Building wheel for kerberos (setup.py): finished with status 'error'
     [exec]   Running setup.py clean for kerberos
     [exec]    command: /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-mz_7t7fa/kerberos/setup.py'"'"'; __file__='"'"'/tmp/pip-install-mz_7t7fa/kerberos/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-fbvdm0s1
     [exec]        cwd: /tmp/pip-install-mz_7t7fa/kerberos/
     [exec]   Complete output (14 lines):
     [exec]   running bdist_wheel
     [exec]   running build
     [exec]   running build_ext
     [exec]   building 'kerberos' extension
     [exec]   creating build
     [exec]   creating build/temp.linux-x86_64-3.6
     [exec]   creating build/temp.linux-x86_64-3.6/src
     [exec]   x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.6m -I/home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/include/python3.6m -c src/base64.c -o build/temp.linux-x86_64-3.6/src/base64.o /bin/sh: 1: krb5-config: not found
     [exec]   x86_64-linux-gnu-gcc: error: /bin/sh:: No such file or directory
     [exec]   x86_64-linux-gnu-gcc: error: 1:: No such file or directory
     [exec]   x86_64-linux-gnu-gcc: error: krb5-config:: No such file or directory
     [exec]   x86_64-linux-gnu-gcc: error: not: No such file or directory
     [exec]   x86_64-linux-gnu-gcc: error: found: No such file or directory
     [exec]   error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
     [exec]   ----------------------------------------
     [exec]   ERROR: Failed building wheel for kerberos
     [exec] Failed to build kerberos
     [exec] Installing collected packages: humanize, docopt, six, parse, parse-type, colorful, tag-expressions, pysingleton, radish-bdd, PyHamcrest, aenum, isodate, kerberos
     [exec]     Running setup.py install for kerberos: started
     [exec]     ERROR: Command errored out with exit status 1:
     [exec]      command: /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-mz_7t7fa/kerberos/setup.py'"'"'; __file__='"'"'/tmp/pip-install-mz_7t7fa/kerberos/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-aljw5xsj/install-record.txt --single-version-externally-managed --compile --install-headers /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/include/site/python3.6/kerberos
     [exec]          cwd: /tmp/pip-install-mz_7t7fa/kerberos/
     [exec]     Complete output (14 lines):
     [exec]     running install
     [exec]     Running setup.py install for kerberos: finished with status 'error'    running build
     [exec] 
     [exec]     running build_ext
     [exec]     building 'kerberos' extension
     [exec]     creating build
     [exec]     creating build/temp.linux-x86_64-3.6
     [exec]     creating build/temp.linux-x86_64-3.6/src
     [exec]     x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.6m -I/home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/include/python3.6m -c src/base64.c -o build/temp.linux-x86_64-3.6/src/base64.o /bin/sh: 1: krb5-config: not found
     [exec]     x86_64-linux-gnu-gcc: error: /bin/sh:: No such file or directory
     [exec]     x86_64-linux-gnu-gcc: error: 1:: No such file or directory
     [exec]     x86_64-linux-gnu-gcc: error: krb5-config:: No such file or directory
     [exec]     x86_64-linux-gnu-gcc: error: not: No such file or directory
     [exec]     x86_64-linux-gnu-gcc: error: found: No such file or directory
     [exec]     error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
     [exec]     ----------------------------------------
     [exec] ERROR: Command errored out with exit status 1: /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-mz_7t7fa/kerberos/setup.py'"'"'; __file__='"'"'/tmp/pip-install-mz_7t7fa/kerberos/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-aljw5xsj/install-record.txt --single-version-externally-managed --compile --install-headers /home/smallette/git/apache/tinkerpop/gremlin-python/target/python3/env/include/site/python3.6/kerberos Check the logs for full command output.
```

Any idea what might be wrong?

**Comment by vtslab** [comment]
Sounds good. Regarding pykerberos, you see it nicely in the docker/Dockerfile in the diff above: you need the libkrb5-dev and krb5-user deb packages locally. That is why end users need to do an additional pip install gremlinpython[kerberos]: you do not want everyone running into this and force them to do these installs.

File: docs/src/reference/gremlin-applications.asciidoc [hunk 1]
```diff
@@ -1076,7 +1076,8 @@ Gremlin-Console |PLAIN SASL (username/password) |3.0.0-incubating
 |Pluggable SASL |3.0.0-incubating
 |GSSAPI SASL (Kerberos) |3.3.0
 |Gremlin.NET |PLAIN SASL |3.3.0
-|Gremlin-Python |PLAIN SASL |3.2.2
+1.2+v|Gremlin-Python |PLAIN SASL |3.2.2
```
Comment (spmallette) [inline docs/src/reference/gremlin-applications.asciidoc:1079]: Did you mean to have that "1.2+v" in there that way? Won't that break the table structure?

**Comment by spmallette** [comment]
I just opened #1290 which has all of these commits plus one of mine that allows Neo4j to build properly. We can close this one now.

File: docs/src/reference/gremlin-variants.asciidoc [hunk 2]
```diff
@@ -642,6 +642,7 @@ To install Gremlin-Python, use Python's link:https://en.wikipedia.org/wiki/Pip_(
 [source,bash]
 ----
 pip install gremlinpython
+pip install gremlinpython[kerberos]     # Optional, not available on Microsoft Windows
```
Comment (spmallette) [inline docs/src/reference/gremlin-variants.asciidoc:645]: I was trying to understand this syntax, specifically the addition of the bracketed "[kerberos]". I see it described pretty nicely here:

https://stackoverflow.com/a/46775606/1831717

that led me to this link:

https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

should "kerberos" be defined in `extra_require"?

**Comment by spmallette** [comment]
Note that I made this addition to the dev docs around those two additional requirements:

https://github.com/apache/tinkerpop/pull/1290/commits/ec206814c822d56b7af0641cd29ddcd2600ac360

Please let me know if that change looks ok to you.

File: docs/src/reference/gremlin-variants.asciidoc (hunk 2 above)
Comment (vtslab) [reply docs/src/reference/gremlin-variants.asciidoc:645]: Sure, in fact it is on line 83 of the setup.py of gremlin-python.

**Comment by vtslab** [comment]
Good point to also update the dev docs.

File: docs/src/reference/gremlin-variants.asciidoc (hunk 2 above)
Comment (spmallette) [reply docs/src/reference/gremlin-variants.asciidoc:645]: sorry - i don't know how i missed that.......

**Pull request closed without merging**